source/state_representation/include/state_representation/IOState.hpp

```diff
@@ -0,0 +1,220 @@
+  /**
+   * @brief Getter of the names
+   * @return The vector of strings containing the IO names
+   */
+  const std::vector<std::string>& get_names() const;
+
+  /**
+   * @brief Get IO index by the name of the IO, if it exists
+   * @param io_name The name of the desired IO
+   * @throws IONotFoundException if the desired IO doesn't exist
+   * @return The index of the IO, if it exists
+   */
+  unsigned int get_io_index(const std::string& io_name) const;
+
+  /**
+   * @brief Returns the values of the IO state as an Eigen vector
+   */
+  Eigen::Vector<T, Eigen::Dynamic> data() const;
+
+  /**
+   * @brief Returns the values of the IO state an Eigen array
+   */
+  Eigen::Array<T, Eigen::Dynamic, 1> array() const;
+
+  /**
+   * @brief Setter of the names from the number of IOs
+   * @param nb_ios The number of IOs of the IO state
+   */
+  void set_names(unsigned int nb_ios);
+
+  /**
+   * @brief Setter of the names from a list of IO names
+   * @param names The vector of strings containing the IO names
+   */
+  void set_names(const std::vector<std::string>& names);
+
+  /**
+   * @brief Set the vlues of the IO state from a single Eigen vector
+   * @param The data vector
+   */
+  void set_data(const Eigen::Vector<T, Eigen::Dynamic>& data);
+
+  /**
+   * @brief Set the values of the IO state from a single std vector
+   * @param The data vector
+   */
+  void set_data(const std::vector<T>& data);
+
+  /**
+   * @brief Check if the IO group is incompatible for operations with the state given as argument
+   * @param state The state to check compatibility with
+   */
+  bool is_incompatible(const State& state) const override;
+
+  /**
+   * @brief Return the IO values as a std vector
+   * @return The IO values as a std vector
+   */
+  std::vector<T> to_std_vector() const;
+
+protected:
+  /**
+   * @brief Empty constructor for an IO state
+   */
+  IOState() = default;
+
+  /**
+   * @brief Constructor with name and number of IOs provided
+   * @param name The name of the associated IO state
+   * @param nb_ios The number of IOs for initialization
+   */
+  explicit IOState(const std::string& name, unsigned int nb_ios);
+
+  /**
+   * @brief Constructor with name and list of IO names provided
+   * @param name The name of the associated IO state
+   * @param io_name List of IO names
+   */
+  IOState(const std::string& name, const std::vector<std::string>& io_names);
+
+  /**
+   * @brief Swap the values of the IO states
+   * @param state1 IO state to be swapped with 2
+   * @param state2 IO state to be swapped with 1
+   */
+  friend void swap(IOState& state1, IOState& state2) {
+    swap(static_cast<State&>(state1), static_cast<State&>(state2));
+    std::swap(state1.names_, state2.names_);
+    std::swap(state1.data_, state2.data_);
+  }
+
+  static void assert_index_in_range(unsigned int io_index, unsigned int size);
+
+  std::vector<std::string> names_;///< names of the IOs
+  Eigen::Vector<T, Eigen::Dynamic> data_;///< IO values
```

Comment:
Any particular reason you are using an Eigen vector as the data container type? I guess for conventions across the other class and enabling the `data()` and `to_std_vector()` methods?

Reply:
Yes indeed, for that reason exactly.